### PR TITLE
Add a session limit to Crashlytics Firelog

### PR DESCRIPTION
**Pull request by jakeouellette**

### CODE REVIEW

File: firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/persistence/CrashlyticsReportPersistence.java [hunk 1]
```diff
@@ -207,17 +207,53 @@ public void finalizeReports(String currentSessionId) {
...
+    Collections.sort(priorityReports, SESSION_ID_COMPARATOR);
+    List<File> reports = getFilesInDirectory(reportsDirectory, filter);
+    Collections.sort(reports, SESSION_ID_COMPARATOR);
+    reportFiles.ensureCapacity(priorityReports.size() + reports.size());
+    reportFiles.addAll(priorityReports);
+    reportFiles.addAll(reports);
+    return reportFiles;
+  }
+
+  public List<CrashlyticsReport> loadFinalizedReports() {
```
Comment (mrwillis21) [inline firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/persistence/CrashlyticsReportPersistence.java]: We should add javadoc pertaining to the expected sort order here.

Comment (mrwillis21) [reply firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/persistence/CrashlyticsReportPersistence.java]: Also, please pull this method up above the private methods.

File: firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/persistence/CrashlyticsReportPersistence.java [hunk 2]
```diff
@@ -207,17 +207,53 @@ public void finalizeReports(String currentSessionId) {
...
+      for (File reportFile : filesToRemove) {
+        reportFile.delete();
+      }
     }
-    final List<File> reports = getAllFilesInDirectory(reportsDirectory);
-    for (File reportFile : reports) {
+  }
+
+  @NonNull
+  private List<File> getAllFinalizedReportFiles() {
```
Comment (mrwillis21) [inline firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/persistence/CrashlyticsReportPersistence.java]: javadoc pertaining to expected sort order would be helpful here.

File: firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/persistence/CrashlyticsReportPersistence.java [hunk 3]
```diff
@@ -207,17 +207,53 @@ public void finalizeReports(String currentSessionId) {
...
+    List<File> reports = getAllFilesInDirectory(reportsDirectory);
+    Collections.sort(reports, SESSION_ID_COMPARATOR);
+    allReportsFiles.ensureCapacity(priorityReports.size() + reports.size());
+    allReportsFiles.addAll(priorityReports);
+    allReportsFiles.addAll(reports);
+    return allReportsFiles;
+  }
+
+  @NonNull
+  private List<File> getFinalizedReportFiles(FilenameFilter filter) {
```
Comment (mrwillis21) [inline firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/persistence/CrashlyticsReportPersistence.java]: As above, javadoc pertaining to the expected sort order would be helpful.

File: firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/persistence/CrashlyticsReportPersistence.java [hunk 4]
```diff
@@ -207,17 +207,53 @@ public void finalizeReports(String currentSessionId) {
...
     }
-    final List<File> reports = getAllFilesInDirectory(reportsDirectory);
-    for (File reportFile : reports) {
+  }
+
+  @NonNull
+  private List<File> getAllFinalizedReportFiles() {
+    final ArrayList<File> allReportsFiles = new ArrayList<>();
+    List<File> priorityReports = getAllFilesInDirectory(priorityReportsDirectory);
+    Collections.sort(priorityReports, SESSION_ID_COMPARATOR);
```
Comment (mrwillis21) [inline firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/persistence/CrashlyticsReportPersistence.java]: How would you feel about reducing the repeated code in these two methods by adding another helper function like so?

    private static List<File> getSortedReportFiles(List<File> priorityReports, List<File> reports) {
        Collections.sort(priorityReports, SESSION_ID_COMPARATOR);
        Collections.sort(reports, SESSION_ID_COMPARATOR);
        final ArrayList<File> allReportsFiles = new ArrayList<>();
        allReportsFiles.ensureCapacity(priorityReports.size() + reports.size());
        allReportsFiles.addAll(priorityReports);
        allReportsFiles.addAll(reports);
        return allReportsFiles;
    }

File: firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/persistence/CrashlyticsReportPersistence.java [hunk 5]
```diff
@@ -207,17 +207,53 @@ public void finalizeReports(String currentSessionId) {
...
   }
 
-  public List<CrashlyticsReport> loadFinalizedReports() {
-    final List<CrashlyticsReport> allReports = new ArrayList<>();
-    final List<File> priorityReports = getAllFilesInDirectory(priorityReportsDirectory);
-    for (File reportFile : priorityReports) {
-      allReports.add(TRANSFORM.reportFromJson(readTextFile(reportFile)));
+  private void capFinalizedReports() {
+    List<File> allReportFiles = getAllFinalizedReportFiles();
+    if (allReportFiles.size() > defaultMaxReportsToKeep) {
```
Comment (mrwillis21) [inline firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/persistence/CrashlyticsReportPersistence.java]: Any chance we could extract a variable for `size` here, outside the `if` to clarify that the same value should be used in the check and in the `subList` call?

Comment (jakeouellette) [reply firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/persistence/CrashlyticsReportPersistence.java]: sure

File: firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/persistence/CrashlyticsReportPersistence.java (hunk 4 above)
Comment (jakeouellette) [reply firebase-crashlytics/src/main/java/com/google/firebase/crashlytics/internal/persistence/CrashlyticsReportPersistence.java]: love it

**Pull request merged**